### PR TITLE
[AMDGPU] Adds pre-commit test for fmul-select combine

### PR DESCRIPTION
**Pull request by vg0204**
This adds the f32/f64/f16/bf16 test cases for below pattern :

`fmul x, select(y, A, B)`
with just one use of select Inst above.

It acts as pre-commit tests for dagCombining above pattern into cheaper ldexp in case of non-inlline 32 bit-constants. (#111109)

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+; GFX1100-LABEL: fmul_select_f32_test1:
+; GFX1100:       ; %bb.0:
+; GFX1100-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX1100-NEXT:    v_and_b32_e32 v1, 1, v1
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1) | instskip(SKIP_1) | instid1(VALU_DEP_1)
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, 1, v1
+; GFX1100-NEXT:    v_cndmask_b32_e64 v1, 1.0, 2.0, vcc_lo
+; GFX1100-NEXT:    v_mul_f32_e32 v0, v0, v1
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %1 = select i1 %bool, float 2.000000e+00, float 1.000000e+00
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:34]: Use named values in tests

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -verify-machineinstrs | FileCheck -check-prefix=GFX9 %s
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 -verify-machineinstrs | FileCheck -check-prefix=GFX1030 %s
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 -verify-machineinstrs | FileCheck -check-prefix=GFX1100 %s
+
+define float @fmul_select_f32_test1(float %x, i1 %bool) {
+; GFX9-LABEL: fmul_select_f32_test1:
+; GFX9:       ; %bb.0:
+; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX9-NEXT:    v_and_b32_e32 v1, 1, v1
+; GFX9-NEXT:    v_cmp_eq_u32_e32 vcc, 1, v1
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:11]: Better to have a real compare as the boolean test value, e.g. use an i32 arg and icmp eq %arg, 0. The non-boolean source i1 values are an edge case with more instructions

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll [hunk 3]
```diff
@@ -0,0 +1,342 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -verify-machineinstrs | FileCheck -check-prefix=GFX9 %s
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 -verify-machineinstrs | FileCheck -check-prefix=GFX1030 %s
+;RUN: llc < %s -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 -verify-machineinstrs | FileCheck -check-prefix=GFX1100 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:4]:
```suggestion
 ;RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s | FileCheck -check-prefix=GFX9 %s
; RUN: llc-mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 < %s | FileCheck -check-prefix=GFX1030 %s
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 < %s | FileCheck -check-prefix=GFX1100 %s
```

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, 1, v5
+; GFX1100-NEXT:    v_dual_mov_b32 v6, v4 :: v_dual_cndmask_b32 v5, 0x3ff00000, v8
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1) | instskip(NEXT) | instid1(VALU_DEP_2)
+; GFX1100-NEXT:    v_mul_f64 v[0:1], v[0:1], v[6:7]
+; GFX1100-NEXT:    v_mul_f64 v[2:3], v[2:3], v[4:5]
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %1 = select <2 x i1> %bool, <2 x double> <double 0.500000e+00, double 0.500000e+00>, <2 x double> <double 1.000000e+00, double 1.000000e+00>
+  %ldexp = fmul <2 x double> %x, %1
+  ret <2 x double> %ldexp
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:342]: Also test the f16 case.

Also negative tests where the constants aren't the right values.

Might also be worth testing cases with all SGPR inputs

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:342]: Also test the negative constants

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1) | instskip(SKIP_1) | instid1(VALU_DEP_1)
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, 1, v2
+; GFX1100-NEXT:    v_cndmask_b32_e64 v2, 1.0, 2.0, vcc_lo
+; GFX1100-NEXT:    v_dual_mul_f32 v0, v0, v2 :: v_dual_and_b32 v3, 1, v3
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1) | instskip(SKIP_1) | instid1(VALU_DEP_1)
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, 1, v3
+; GFX1100-NEXT:    v_cndmask_b32_e64 v3, 1.0, 2.0, vcc_lo
+; GFX1100-NEXT:    v_mul_f32_e32 v1, v1, v3
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %1 = select <2 x i1> %bool, <2 x float> <float 2.000000e+00, float 2.000000e+00>, <2 x float> <float 1.000000e+00, float 1.000000e+00>
```
Comment (vg0204) [inline llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll:112]: @arsenm, can you let me know, how to check for vector SDNodes to handle such cases, where each value of const vector SDNode need to be insepected.

File: llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll (hunk 1 above)
Comment (vg0204) [reply llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll]: I guess need to do this manually, as there exists currently no way to automate it?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/combine-fmul-sel.ll]: opt -S -passes=instnamer

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GFX1100-NEXT:    v_mul_f32_e32 v0, v0, v1
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq i32 %bool.arg1, %bool.arg2
+  %1 = select i1 %bool, float 0.500000e+00, float 1.000000e+00
+  %ldexp = fmul float %x, %1
+  ret float %ldexp
+}
+
+define <2 x float> @fmul_select_f32_test3(<2 x float> %x, <2 x i32> %bool.arg1, <2 x i32> %bool.arg2) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:68]:
```suggestion
define <2 x float> @fmul_select_v2f32_test3(<2 x float> %x, <2 x i32> %bool.arg1, <2 x i32> %bool.arg2) {
```

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 7]
```diff
@@ -0,0 +1,781 @@
...
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, v2, v4
+; GFX1100-NEXT:    v_cndmask_b32_e64 v2, 1.0, 0.5, vcc_lo
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, v3, v5
+; GFX1100-NEXT:    v_cndmask_b32_e64 v3, 1.0, 0.5, vcc_lo
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GFX1100-NEXT:    v_dual_mul_f32 v0, v0, v2 :: v_dual_mul_f32 v1, v1, v3
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq <2 x i32> %bool.arg1, %bool.arg2
+  %1 = select <2 x i1> %bool, <2 x float> <float 0.500000e+00, float 0.500000e+00>, <2 x float> <float 1.000000e+00, float 1.000000e+00>
+  %ldexp = fmul <2 x float> %x, %1
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:142]:
```suggestion
  %select = select <2 x i1> %bool, <2 x float> <float 0.500000e+00, float 0.500000e+00>, <2 x float> <float 1.000000e+00, float 1.000000e+00>
  %ldexp = fmul <2 x float> %x, %select
```

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 8]
```diff
@@ -0,0 +1,781 @@
...
+; GFX1100-NEXT:    v_cndmask_b32_e32 v1, 0x3c00, v5, vcc_lo
+; GFX1100-NEXT:    v_pack_b32_f16 v1, v1, v2
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GFX1100-NEXT:    v_pk_mul_f16 v0, v0, v1
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq <2 x i32> %bool.arg1, %bool.arg2
+  %1 = select <2 x i1> %bool, <2 x half> <half 0.500000e+00, half 0.500000e+00>, <2 x half> <half 1.000000e+00, half 1.000000e+00>
+  %ldexp = fmul <2 x half> %x, %1
+  ret <2 x half> %ldexp
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:781]: Test 0 / -0. Some cases with mismatched signs on each side. Negative test for multiple uses

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 9]
```diff
@@ -0,0 +1,1191 @@
...
+; GFX1100-NEXT:    v_mul_f32_e32 v0, v0, v1
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq i32 %bool.arg1, %bool.arg2
+  %y = select i1 %bool, float -0.000000e+00, float 0.000000e+00
+  %ldexp = fmul float %x, %y
+  ret float %ldexp
+}
+
+
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:343]:
```suggestion
```

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 10]
```diff
@@ -0,0 +1,1191 @@
...
+; GFX1100-NEXT:    v_mov_b32_e32 v3, 0xcc00
+; GFX1100-NEXT:    v_cmp_eq_u32_e32 vcc_lo, v1, v2
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_2) | instskip(NEXT) | instid1(VALU_DEP_1)
+; GFX1100-NEXT:    v_cndmask_b32_e32 v1, 0xd000, v3, vcc_lo
+; GFX1100-NEXT:    v_mul_f16_e32 v0, v0, v1
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq i32 %bool.arg1, %bool.arg2
+  %y = select i1 %bool, half -1.600000e+01, half -3.200000e+01
+  %ldexp = fmul half %x, %y
+  ret half %ldexp
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:1190]: Add a few bfloat cases

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 11]
```diff
@@ -0,0 +1,1191 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+;RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s | FileCheck -check-prefix=GFX9 %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:2]: Test gfx700

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 12]
```diff
@@ -0,0 +1,1191 @@
...
+;RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1030 < %s | FileCheck -check-prefix=GFX1030 %s
+;RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 < %s | FileCheck -check-prefix=GFX1100 %s
+
+define float @fmul_select_f32_test1(float %x, i32 %bool.arg1, i32 %bool.arg2) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:6]: Add a couple cases with inreg arguments for the scalar case

**Review by arsenm — COMMENTED** [review]
almost there, but few more cases might be useful

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll (hunk 10 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:1190]: Also vectors

**Comment by arsenm** [comment]
Test really high powers of 2. Particularly the cases where the exponent is > 64 (or < -16) and thus no longer an inline immediate

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll (hunk 12 above)
Comment (vg0204) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:6]: I did not understand what you mean? Can you give one very small example!

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:6]: ```define float @s_fmul_select_f32_test1(float inreg %x, i32 inreg %bool.arg1, i32 inreg %bool.arg2) {```

(Could also replace bool.arg2 with a constant everywhere)

**Comment by vg0204** [comment]
> Test really high powers of 2. Particularly the cases where the exponent is > 64 (or < -16) and thus no longer an inline immediate

```
define float @fmul_select_f32_test11(float %x, i32 %bool.arg1, i32 %bool.arg2) {
  %bool = icmp eq i32 %bool.arg1, %bool.arg2
  %y = select i1 %bool, float -3.0223145e+23, float -7.2057594e+16
  %ldexp = fmul float %x, %y
  ret float %ldexp
}
```
Note :  **_-3.0223145e+23 = -2^78_**  and  _**-7.2057594e+16 = -2^56**_
Whats wrong with this as it is giving this error?

```
/opt/compiler-explorer/clang-trunk/bin/llc: error: /opt/compiler-explorer/clang-trunk/bin/llc: <source>:72:31: error: floating point constant invalid for type
  %y = select i1 %bool, float -3.0223145e+23, float -7.2057594e+16
                              ^
Compiler returned: 1
```

**Comment by arsenm** [comment]
> > Test really high powers of 2. Particularly the cases where the exponent is > 64 (or < -16) and thus no longer an inline immediate
> 
> ```
> define float @fmul_select_f32_test11(float %x, i32 %bool.arg1, i32 %bool.arg2) {
>   %bool = icmp eq i32 %bool.arg1, %bool.arg2
>   %y = select i1 %bool, float -3.0223145e+23, float -7.2057594e+16
>   %ldexp = fmul float %x, %y
>   ret float %ldexp
> }
> ```
> 
> Note : **_-3.0223145e+23 = -2^78_** and _**-7.2057594e+16 = -2^56**_ Whats wrong with this as it is giving this error?
> 
> ```
> /opt/compiler-explorer/clang-trunk/bin/llc: error: /opt/compiler-explorer/clang-trunk/bin/llc: <source>:72:31: error: floating point constant invalid for type
>   %y = select i1 %bool, float -3.0223145e+23, float -7.2057594e+16

You can only use the decimal notation for cases that are exactly representable in decimal. You need to use the standard hex representation (I really wish the IR would switch to using C99 hex float notation)

**Comment by arsenm** [comment]
You can always write something simpler and let constant folding show you how it should be printed: https://godbolt.org/z/PsdMnq37T

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll (hunk 12 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:7]: Didn't add the scalar cases?

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll [hunk 13]
```diff
@@ -0,0 +1,2719 @@
...
+; GFX1100-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GFX1100-NEXT:    v_lshrrev_b32_e32 v0, 16, v0
+; GFX1100-NEXT:    s_setpc_b64 s[30:31]
+  %bool = icmp eq i32 %bool.arg1, %bool.arg2
+  %y = select i1 %bool, bfloat 0xRE000, bfloat 0xRDB80
+  %ldexp = fmul bfloat %x, %y
+  ret bfloat %ldexp
+}
+
+define bfloat @fmul_select_bf16_test11(bfloat %x, i32 %bool.arg1, i32 %bool.arg2) {
```
[the remark this answers is not in the document]
Comment (vg0204) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:2645]: > Can you name these to state the values, it's hard to read the hex values

Sure thing!

**Comment by vg0204** [comment]
Kind Ping!

File: llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll (hunk 12 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/dagcombine-fmul-sel.ll:7]:
```
define float @v_s_fmul_select_f32_test1(float inreg %x, i32 inreg %bool.arg1) {
  %bool = icmp eq i32 %bool.arg1, 0
  %y = select i1 %bool, float 2.000000e+00, float 1.000000e+00
  %ldexp = fmul float %x, %y
  ret float %ldexp
}

; with scalar output
define amdgpu_ps i32 @s_fmul_select_f32_test1(float inreg %x, i32 inreg %bool.arg1) {
  %bool = icmp eq i32 %bool.arg1, 0
  %y = select i1 %bool, float 2.000000e+00, float 1.000000e+00
  %ldexp = fmul float %x, %y
  %cast = bitcast float %ldexp to i32
  ret i32 %cast
}
```

**Comment by vg0204** [comment]
Ping @arsenm!

**Review by arsenm — APPROVED** [review]

**Pull request merged**